design/2025-10-persistent-replica-runtime.md

```diff
@@ -0,0 +1,258 @@
+associated with. They are not associated with the in-memory instance but the
+activation of the protection.
+
+### Outcome replica (effect handler)
+
+An outcome replica is a special kind of replica that records whether actions
+have been issued for intents. The semantic meaning of the outcome is not
+relevant, e.g. failure to ban a user is still an outcome that needs to be
+recorded and is the outcome of the intent.
+
+outcome replicas for protections are partitioned in the same way as the intent
+replicas, it's just the capability providers are responsible for producing the
+actions.
+
+## Semantics
+
+### Actor migration
+
+Actors need to be transportable across node.js worker threads in order to
+balance CPU.
+
+### Preemption
+
+replica actor client code will be interruptive in order to gracefully handle
+shutdown and migration of actors.
+
+This will be achieved by abusing I/O operations to signal interrupts. CPU bound
+actor code will still have to cooperatively check for an interrupt signal.
+
+### Garbage collection
+
+Replica dependents will formally be tracked. Transitive dependencies will also
+need to be considered for rollbacks to be successful. replicas will only be able
+to discard an old revision when all transitive dependents have successfully
+migrated to a superseding revision.
+
+### Scheduling
+
+The scheduler is responsible for starting replica actors and migrating them when
+CPU becomes exhausted in worker threads. The scheduler gives each actor a
+controller that signals interrupts. The store is expected to also receive the
+controller so that it can preempt on I/O.
+
+The scheduler will have to be communicated `os.loadavg()` from each worker
+thread and use this to trigger actor migration. Realistically we can probably
+only move actors from high utilisation to low utilisation and have some strategy
+that tries to identify and isolate the cpu intensive actors for reporting.
+
+### Crashing
+
+What happens when an actor crashes?
+
+#### Poisoned delta
+
+If the revision reducer crashes when processing a delta, and continues to crash
+when re-attempted, the delta will be skipped with an explicit acknowledgement.
+This signifies a bug in the implementation of the revision reducer and is really
+a critical issue. It's important not to bring down the whole system, but quite
+honestly the existence of a reducer bug in critical areas can be as bad as that.
+Protections obviously can have faulty reducers and don't effect other
+functionality.
+
+#### Other issues
+
+The actor should be restarted with exponential backoff.
+
+### Messaging
+
+How do deltas get communicated to dependencies? especially if the in-memory
+actors can crash and be unreliable for message delivery.
+
+The deltas get written to the persistent store and a message is sent to the
+other threads about the write so that the delta can be read back from the store
+and propagated to dependencies.
```

Comment:
After delta-write, we can just push the entire delta to the dependencies as part of the hint message. The delta has provancne data so each dependency will know if it has missed a delta from that and check. This saves us a little bit of latency.